x16-smc.ino

```diff
@@ -572,13 +641,35 @@ void mouseClockIrq() {
 // ----------------------------------------------------------------
 // Bootloader Startup
 // ----------------------------------------------------------------
-void startBootloader() {
-  if (bootloaderFlags == 3 && bootloaderTimer > 0) {
-    ((void(*)(void))BOOTLOADER_START_ADDR)();
-  }
-  else {
-    bootloaderTimer = 50;
+void evaluateBootloaderKeypress() {
+  if (bootloaderTimer > 0 && specialModeActive != 1) {
+    uint8_t temp = SREG;
+    cli();
+    bootloaderTimer--;
+    SREG = temp;
+
+    uint8_t currentKeysPressed = digitalRead_opt(POWER_BUTTON_PIN) | (digitalRead_opt(RESET_BUTTON_PIN) << 1) | (digitalRead_opt(NMI_BUTTON_PIN) << 2);
+    currentKeysPressed ^= 7; // Buttons are active low. Flip bits, to let the following logic use active high logic instead (more intuitive)
+
+    // Reduce timer to 0.5 s once the first key is pressed down
+    if (currentKeysPressed && bootloaderTimer > 50) {
+      bootloaderTimer = 50;
+    }
+
+    // specialModeRequested == 0 -> try to start bootloader (Power(0x01) + Reset(0x02) + no NMI)
+    if (specialModeRequested == 0 && currentKeysPressed == 3) { // Power (1) + Reset (2) + no NMI
+      cli();
+      ((void(*)(void))BOOTLOADER_START_ADDR)(); // Start bootloader
+    }
+
+    // specialModeRequested == 1 -> try to enable flash programming (Power(0x01) + NMI(0x04) + no Reset)
+    if (specialModeRequested == 1 && currentKeysPressed == 5) { // Power (1) + NMI (4) + no Reset
+      specialModeActive = 1;
+      // The user is currently pressing power+NMI. We want to disable key press until programming is done.
+      bootloaderTimer = 50; // Will not be decremented
+    }
```

Comment:
Good point. I agree that we should make it usable on machines without a dedicated NMI button if that is a thing. My thought was to prevent starting this feature accidentally (or maliciously, maybe I am a bit too paranoid). I can try to simply change buttons to power+reset on this mode as well.

How does this require you to hold buttons longer? Have you experienced this, or is it based on something you read in the code?

Reply:
It is a bit different. The current code doesn't really require both buttons to be pressed at the same time. It only requires them to be pressed/released within the 0.5 s threshold. I managed to press/release them in a way that the new code didn't accept it.

I don't know if it needs to be changed. It's more of an observation.